### PR TITLE
fix: deadlock issue

### PR DESCRIPTION
**Pull request by amandavialva01**
## Ticket
DET-10441

## Description
Fix deadlocking issue with creating or verifying existence of a kubernetes namespace 

## Checklist

- [ ] Changes have been manually QA'd
- [ ] New features have been approved by the corresponding PM
- [ ] User-facing API changes have the "User-facing API Change" label
- [ ] Release notes have been added as a separate file under `docs/release-notes/`
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses have been included for new code which was copied and/or modified from any external code

[DET-123]: https://hpe-aiatscale.atlassian.net/browse/DET-123?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by davidfluck-hpe** [comment]
Were you able to reproduce the issue?

**Comment by amandavialva01** [comment]
> Were you able to reproduce the issue?

This PR seems to resolve the issue, which I have tested by helm installing Determined with this commit SHA onto a remote cluster and running the workflow that produces the deadlock when a SHA from main is used with the helm install instead.
Working on end-to-end tests that I'll add to this PR tomorrow

**Comment by davidfluck-hpe** [comment]
Fantastic, thank you! What is the workflow that reproduces the issue?

**Comment by amandavialva01** [comment]
> Fantastic, thank you! What is the workflow that reproduces the issue?

If you're running a remote GKE cluster: Create a workspace bound to an auto-created namespace and start a task (I used a JL instance) in that workspace. Then, kill the master deployment pod and wait for the k8s API server to recreate it. Then, in open the "Edit Workspace" modal in the webUI and save the workspace. You'll see the loading sign, and then nothing will load after that 🙈😂

File: master/internal/rm/kubernetesrm/jobs.go [hunk 1]
```diff
@@ -2173,11 +2174,9 @@ func (j *jobsService) verifyNamespaceExists(namespace string) error {
 
 	// Since we don't know whether the namespace exists yet, and we don't want to do duplicate
 	// namespace informers, add all non-auto-created namespaces bound to a workspace with a map.
```
Comment (kkunapuli) [inline master/internal/rm/kubernetesrm/jobs.go:2176]: Would it be simpler to check if an informer exists before calling `syncNamespaces` in `createNamespace`? (e.g., using the logic `if _, ok := j.namespacesWithInformers[namespace]; !ok {` )

Depending on the intent of `syncNamespaces`, that check could be added to the beginning of the function instead. So, before `syncNamespaces` creates any informers, it confirms that the informers don't exist yet.

Comment (amandavialva01) [reply master/internal/rm/kubernetesrm/jobs.go:2176]: That logic is added in line 257 to the `syncNamespaces` function! I refactored `syncNamespaces` so that instead of checking the `namespacesWithInformers` map before calling `syncNamespaces` each time, we check the `namespacesWithInformers` map in `syncNamespaces` to avoid missing any map updates

Comment (kkunapuli) [reply master/internal/rm/kubernetesrm/jobs.go:2176]: Oh, I understand now - thank you for explaining! I see how `syncNamespaces` checks for existing informers before creating new ones and the other changes are to accept a single namespace instead of a slice of namespaces.

File: master/internal/rm/kubernetesrm/jobs.go [hunk 2]
```diff
@@ -253,18 +253,22 @@ func newJobsService(
...
-	err = j.startPreemptionListeners(ns)
-	if err != nil {
-		return err
-	}
-
 	for _, namespace := range ns {
+		if _, ok := j.namespacesWithInformers[namespace]; ok {
+			continue
+		}
 		j.namespacesWithInformers[namespace] = true
```
Comment (kkunapuli) [inline master/internal/rm/kubernetesrm/jobs.go:260]: If the `startEventListeners` below fails, the namespace wouldn't have an informer yet though, right? Should we only set `namespacesWithInformers` to true if no error is returned?

**Review by kkunapuli — APPROVED** [review]
LGTM! 

I just have a question about when we set `j.namespacesWithInformers[namespace] = true` (line 260).

File: master/internal/rm/kubernetesrm/jobs.go (hunk 1 above)
Comment (kkunapuli) [reply master/internal/rm/kubernetesrm/jobs.go:2176]: For what it's worth, I love this comment. Maybe we could just move it to `syncNamespaces`? I think it's a subtle but important point.

File: master/internal/rm/kubernetesrm/jobs.go (hunk 2 above)
Comment (amandavialva01) [reply master/internal/rm/kubernetesrm/jobs.go:260]: Ooh this is a great point! Yepp I'll make this change

File: master/internal/rm/kubernetesrm/jobs.go (hunk 1 above)
Comment (amandavialva01) [reply master/internal/rm/kubernetesrm/jobs.go:2176]: Good idea, will do!

File: e2e_tests/tests/cluster/test_wksp_nmsp_concurrency.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+from tests import config as conf
+from tests.task import task
+
+
+# Workspace-namespace binding requests were causing a deadlock in our Kubernetes jobs handler when
+# a thread that locks the jobsService tries to reacquire the lock during the execution of a callback
+# function that only gets called when a job is running in the namespace that we want to bind to the
+# workspace. Verify that we don't run into this deadlock when triggering multiple calls to the
+# API handler that binds a workspace to an auto-created namespace, as this request can trigger the
+# deadlock if the namespace (or verify the existence of for the first time) is running a job.
```
Comment (kkunapuli) [inline e2e_tests/tests/cluster/test_wksp_nmsp_concurrency.py:18]: I love this explanation! Great comment and great test.

File: .circleci/devcluster/single-k8s.devcluster.yaml [hunk 4]
```diff
@@ -35,9 +35,3 @@ stages:
           kubeconfig_path: ~/.kube/config
           determined_master_ip: $DOCKER_LOCALHOST
           determined_master_port: 8080
-          internal_task_gateway:
```
Comment (kkunapuli) [inline .circleci/devcluster/single-k8s.devcluster.yaml:38]: Out of curiosity, why is this deleted? Was it a bit of cleanup discovered during testing?

**Review by kkunapuli — APPROVED** [review]
LGTM! Nice work.

File: .circleci/devcluster/single-k8s.devcluster.yaml (hunk 4 above)
Comment (amandavialva01) [reply .circleci/devcluster/single-k8s.devcluster.yaml:38]: Yeaa, I think we only need to use a gateway for mutliRM to make tasks that create resources in the remote cluster (the cluster tied to the additional RM) accessible to the determined master, so single RM clusters don't need that config!
